### PR TITLE
fix prototype datasets data loading tests

### PR DESCRIPTION
**Pull request by pmeier**
This re-enables the serialization test after pytorch/pytorch#74958. Test will fail for next ~3 hours until the other PR is part of the PyTorch core nightlies.

Edit: The PR evolved to something more. Instead of only serialization, we now also test traversability as well loading through a `DataLoader`. Together with #5723 we should now have sufficient coverage for our datasets to not accidentally break data loading workflow.

### CODE REVIEW

File: .circleci/config.yml.in
Comment (pmeier) [inline .circleci/config.yml.in:159]: For some time now `iopath` is no longer a dependency at import and is only used at runtime in

https://github.com/pytorch/vision/blob/93104c1697384273a75381d57be2f872265eeefb/torchvision/prototype/datasets/utils/_internal.py#L175-L183

To my best knowledge, we are currently not using that. Thus, it should be safe to remove here.

---

I was only looking at this, because I was adding the `dill` installation. LMK, if I should bring a separate PR for this change.

File: test/test_prototype_builtin_datasets.py
Comment (pmeier) [inline test/test_prototype_builtin_datasets.py:116]: @ejguan @NivekT since the serialization backend is automatically selected based on the presence of `dill`, it is impossible to test pickle serialization without patching this. For now we only need to patch a single module, but as https://github.com/pytorch/pytorch/pull/74958#issuecomment-1084637227 implies, we need to do this in multiple places in the future. 

Would it be possible give users the option to set the serialization backend? The default can still be `dill` if available otherwise `pickle`.

**Comment by pmeier** [comment]
Even with the patch in https://github.com/pytorch/vision/pull/5711#discussion_r839678058, the `split="val"` configuration [fails](https://app.circleci.com/pipelines/github/pytorch/vision/16234/workflows/2110142b-de36-4ba9-8eef-6997b5cad694/jobs/1315540/tests) due to https://github.com/pytorch/pytorch/pull/74958#issuecomment-1084637227.

File: test/test_prototype_builtin_datasets.py
Comment (ejguan) [reply test/test_prototype_builtin_datasets.py:116]: I see. I can confirm that we will rely on this `DILL_AVAILABLE` for any place in TorchData project to determine if `dill` is available or not.

> Would it be possible give users the option to set the serialization backend? The default can still be dill if available otherwise pickle.

It's doable, but I am not sure if we want to do so because the goal of automatically using `dill` is to reduce the work users need to figure out if the DataPipe is serializable with lambda function.

Comment (pmeier) [reply test/test_prototype_builtin_datasets.py:116]: > I see. I can confirm that we will rely on this `DILL_AVAILABLE` for any place in TorchData project to determine if `dill` is available or not.

The problem is that we will need to patch every single module where this is imported. You cannot patch the place where it is defined, but rather where it is used. If you look above, we are not patching `._utils.serialization` but rather `.datapipes.datapipe`, because this is where the flag is used. If we now need use this flag in multiple modules, we need to patch all of them. This is very brittle.

> It's doable, but I am not sure if we want to do so because the goal of automatically using `dill` is to reduce the work users need to figure out if the DataPipe is serializable with lambda function.

Not sure I understand. If we just keep the same detection as we have now, users that don't care should not see any difference. If `dill` is available, it will be picked up and otherwise `pickle` will be used. But it would give users the option to enforce a particular backend if they need to. Without this option, the environment you use has an effect on the functionality and there is no way change that. I don't think this is good design. 

Even if you don't do it for the users, think about how you want to test `pickle` vs `dill` yourself. Right now the only option is to have two separate workflows one with `dill` installed and one without.

Comment (ejguan) [reply test/test_prototype_builtin_datasets.py:116]: > The problem is that we will need to patch every single module where this is imported. You cannot patch the place where it is defined, but rather where it is used. If you look above, we are not patching `._utils.serialization` but rather `.datapipes.datapipe`, because this is where the flag is used. If we now need use this flag in multiple modules, we need to patch all of them. This is very brittle.

You still can add following code to override the method rather than using `patch`.
```
def state_fn(self):
    return self.__dict__
IterDataPipe. set_getstate_hook(state_fn)
```

> But it would give users the option to enforce a particular backend if they need to. Without this option, the environment you use has an effect on the functionality and there is no way change that. I don't think this is good design.

This is actually a good argument for users who have fully understanding about what they want to achieve. We may be able to expose an API to switch backend if needed, similar to the `set_getstate_hook` but with syntax sugar.

Comment (ejguan) [reply test/test_prototype_builtin_datasets.py:116]: Adding an issue https://github.com/pytorch/data/issues/341

Comment (ejguan) [reply test/test_prototype_builtin_datasets.py:116]: cc: @NivekT Overriding `set_getstate_hook` with the above function won't actually work for all DataPipe like `Forker` https://github.com/pytorch/pytorch/blob/835cc66e5dd26db558931b4fe47b45e08a3a09f7/torch/utils/data/datapipes/iter/combining.py#L158-L167

Then, we should definitely support backend switch that Philip suggested

File: test/test_prototype_builtin_datasets.py
Comment (pmeier) [inline test/test_prototype_builtin_datasets.py:127]: Is it sufficient to test only 1 worker for multi-processing here?

**Review by pmeier — COMMENTED** [review]
There are two distinct failures that will be eliminated by pytorch/pytorch#75034:

- `traverse` is not guarded against infinite recursion loops
- `split="val"` of ImageNet contains a `Demultiplexer`, which is also not guarded against infinite recursion loops

Instead of `xfail`'ing the tests now let's just wait for the other PR to get merged.

**Comment by pmeier** [comment]
This PR is a sister to #5723, which adds a data loader test for DDP.

File: test/test_prototype_builtin_datasets.py
Comment (pmeier) [inline test/test_prototype_builtin_datasets.py:44]:
```suggestion
@pytest.mark.filterwarnings("error")
```

File: test/test_prototype_builtin_datasets.py
Comment (ejguan) [reply test/test_prototype_builtin_datasets.py:127]: Yes

File: test/test_prototype_builtin_datasets.py
Comment (NicolasHug) [inline test/test_prototype_builtin_datasets.py:118]: Should we parametrize over `only_datapipe` or would this be overkill?

File: test/test_prototype_builtin_datasets.py
Comment (NicolasHug) [inline test/test_prototype_builtin_datasets.py:141]: In other tests we do `next(iter(dl))`. Any reason to exhaust the dataloader here?

**Review by NicolasHug — APPROVED** [review]
Thanks @pmeier , minor questions but LGTM.

File: test/test_prototype_builtin_datasets.py
Comment (pmeier) [reply test/test_prototype_builtin_datasets.py:118]: @NicolasHug you are right. Let's be safe here.

@ejguan The test are already failing until your PR is merged so we need to wait anyway.

File: test/test_prototype_builtin_datasets.py
Comment (pmeier) [reply test/test_prototype_builtin_datasets.py:141]: I wasn't sure if just drawing one sample is enough to hit everything. If it is, I'm happy to adapt your suggestion.

**Comment by pmeier** [comment]
[CI is happy](https://app.circleci.com/pipelines/github/pytorch/vision/16404/workflows/5e569d0a-7edf-458c-96f7-1b10848e5a85/jobs/1330310) with today's `torch` / `torchdata` nightlies.

**Pull request merged**
